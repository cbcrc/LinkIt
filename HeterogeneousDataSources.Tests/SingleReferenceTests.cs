using System.Collections.Generic;
using System.Linq;
using ApprovalTests.Reporters;
using NUnit.Framework;
using RC.Testing;

namespace HeterogeneousDataSources.Tests {
    [UseReporter(typeof(DiffReporter))]
    [TestFixture]
    public class SingleReferenceTests
    {
        [Test]
        public void LoadLink_SingleReference()
        {
            var loadLinkExpressions = new List<LoadLinkExpression<SingleReferenceLinkedSource, Image, string>>{
                    new LoadLinkExpression<SingleReferenceLinkedSource,Image, string>(
                        linkedSource => linkedSource.Model.SummaryImageId,
                        (linkedSource, reference) => linkedSource.SummaryImage = reference
                    )
                };

            var content = new SingleReferenceContent{
                Id = 1,
                SummaryImageId = "a"
            };
            var contentLinkedSource = new SingleReferenceLinkedSource(content);

            var asLoadExpressions = loadLinkExpressions
                .Cast<ILoadExpression<SingleReferenceLinkedSource,string>>()
                .ToList();

            var loader = new Loader();
            var dataContext = loader.Load(contentLinkedSource, asLoadExpressions, TestUtil.ImageReferenceTypeConfig);

            var contentLinkedSourceLinkExpressions = loadLinkExpressions
                .Cast<ILinkExpression<SingleReferenceLinkedSource>>()
                .ToList();

            var linker = new Linker();
            linker.Link(dataContext, contentLinkedSource, contentLinkedSourceLinkExpressions);
            ApprovalsExt.VerifyPublicProperties(contentLinkedSource);
        }
    }


    public class SingleReferenceLinkedSource {
        //stle: two steps loading sucks!
        public SingleReferenceLinkedSource(SingleReferenceContent model)
        {
            Model = model;
        }

        public SingleReferenceContent Model { get; private set; }
        public Image SummaryImage;
    }

    public class SingleReferenceContent {
        public int Id { get; set; }
        public string SummaryImageId { get; set; }
    }
}
