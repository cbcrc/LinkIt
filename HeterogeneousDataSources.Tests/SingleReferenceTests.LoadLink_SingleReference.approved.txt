{
  "summaryImage": {
    "id": "a",
    "alt": "alt-a"
  },
  "model": {
    "id": 1,
    "summaryImageId": "a"
  }
}